docs/proposals/202005_query_logging_rate_limiting.md

```diff
@@ -0,0 +1,126 @@
+---
+title: Query Logging and Rate Limiting
+type: proposal
+menu: proposals
+status: unknown
+owner: yashrsharma44, kakkoyun, bwplotka
+---
+
+# Todo
+
+
+### Related Tickets
+
+* Add Query Logging : https://github.com/thanos-io/thanos/issues/170
+* Add generic layer of limits : https://github.com/thanos-io/thanos/issues/2527
+
+## Summary
+
+This proposal describes configuring a new internal feature : Query logging and rate limiting, for internal APIs of Thanos. 
+
+We will go through the problem statement, use-cases and potential solution for the same.
+
+## Problem Statement / Motivation
+
+Conceptually and at a very high level, Thanos is a durable and cheap database, storing a very large amount of metric data in external object storage. This has the possible implication of having expensive data queries if the amount of data required to be queried is large. 
+
+Even after local optimizations and efficient indexing, still querying a large amount of data is very resource-intensive, and unless the user knows what he/she is doing, it can be very expensive. A possible solution apart from the existing ones is limiting the queries based on the time_range and data consumption. 
+
+It does not make sense if the user requests years of data each containing millions of series, and it does not fit in the specification of the local Thanos Store gateway. Getting such a huge amount of data is not pragmatic, and some sort of checks for the user in getting estimates about the kind of requests he/she is making really helps in avoiding such blockhole.
+
+Another feature lacking from Thanos, is tracking/logging of active queries run in Thanos. This is really helpful when a Thanos component goes down, and it would help the administrator/developer in reading the last run queries before the component gets OOM killed. A similar feature exists in Prometheus which was introduced in version 2.16.0. The same feature in Thanos would help in debugging the potential queries that caused the issue.
+
+## Use Cases
+
+* Controlling and limiting the amount of requests made by the internal APIs. 
+* Query logging part would help in debugging the internal requests made.
+* This provides ample of insights about possible bottle-neck of individual components of Thanos.
+
+
+## Goals of this design
+
+* Implement Query Logging for internal APIs of Thanos, StoreAPI in particular
+* The Query Logging to implement a separate logger derived from the base logger. 
+* Implement Rate limiting and tracking of queries of the internal APIs
+
+
+## Proposal
+
+### Query Logging in Thanos
+
+#### Description of the Problem
+
+The Query Logging is a feature which helps in tracking all the active queries made in a database, and helps the users in knowing/debugging what queries made were resource-intensive, and how they can use the details to further optimize the database, or from the user’s perspective, to draw out the line from making such a request.
```

Comment:
Did not we explain the problem already in motivation? :thinking:

Reply:
> Yup, revamped the structure for removing redundant content.